packages/app/src/components/inspector/PatternInspector.tsx

```diff
@@ -0,0 +1,238 @@
+    theme.colors.primary.replace('rgb', 'rgba').slice(0, -1) + ',0.5)';
+
+  const gradientOverlay = `linear-gradient(0deg, ${dragOverlayColor}, ${dragOverlayColor})`;
+  return {
+    display: 'flex',
+    flex: 1,
+    position: 'relative',
+    outline: 'none',
+    borderRadius: '4px',
+    minHeight: '150px',
+    background: (isDragOver ? `${gradientOverlay},` : '') + background,
+    border: (isDragOver ? '2px ' : '0px ') + theme.colors.primaryDark,
+    backgroundColor: 'white',
+    backgroundPosition: 'center',
+    backgroundRepeat: repeat ? 'auto' : 'no-repeat',
+    backgroundSize,
+    imageRendering: 'crisp-edges',
+    width: '100%',
+    justifyContent: 'center',
+  };
+});
+
+const UploadButton = styled.button<{ show: boolean }>(
+  ({ theme, show = false }) => ({
+    color: theme.colors.button.secondaryText,
+    position: 'absolute',
+    background: 'rgba(0, 0, 0, 0.75)',
+    cursor: 'pointer',
+    userSelect: 'auto',
+    alignSelf: 'center',
+    height: '30px',
+    width: '105px',
+    borderRadius: '24px',
+    border: 'none',
+    display: show ? 'block' : 'none',
+    zIndex: 1,
+  }),
+);
+
+interface Props {
+  id: string;
+  images: FileMap;
+  pattern: SketchPattern;
+  onChangeImage?: (image: Sketch.FileRef | Sketch.DataRef) => void;
+  onChangeFillType?: (amount: Sketch.PatternFillType) => void;
+  onChangeTileScale?: (amount: number) => void;
+  createImage: (image: ArrayBuffer, _ref: string) => void;
+}
+
+export type PatternFillTypes = 'Stretch' | 'Fill' | 'Fit' | 'Tile';
+export const patternFillTypeOptions: PatternFillTypes[] = [
+  'Tile',
+  'Fill',
+  'Stretch',
+  'Fit',
+];
+
+export default memo(function PatternInspector({
+  id,
+  images,
+  pattern,
+  onChangeImage,
+  onChangeFillType,
+  onChangeTileScale,
+  createImage,
+}: Props) {
+  const [showButton, setShowButton] = useState(false);
+  const [isDragOver, setIsDragOver] = useState(false);
+
+  const patternType = pattern.patternFillType;
+  const isTile = patternType === Sketch.PatternFillType.Tile;
+
+  const changeFillType = useCallback(
+    (value: PatternFillTypes) => {
+      onChangeFillType?.(Sketch.PatternFillType[value]);
+    },
+    [onChangeFillType],
+  );
+
+  const onSubmitTileScale = useCallback(
+    (value: number) => {
+      onChangeTileScale?.(value / 100);
+    },
+    [onChangeTileScale],
+  );
+
+  const onNudgeTileScale = useCallback(
+    (value: number) => {
+      onChangeTileScale?.(value / 100);
+    },
+    [onChangeTileScale],
+  );
+
+  const background = useMemo(
+    () => getPatternBackground(images, pattern.image),
+    [images, pattern.image],
+  );
+
+  const backgroundSize = useMemo(
+    () => getPatternSize(pattern.patternFillType, pattern.patternTileScale),
+    [pattern.patternFillType, pattern.patternTileScale],
+  );
+
+  const handleImageFile = useCallback(
+    async (file: File) => {
+      const data = await file.arrayBuffer();
+
+      const extension = file.type.split('/')[1];
```

Comment:
We'll probably want a mapping from type => extension, both to show an error for invalid file types, and to handle jpeg. E.g.

```ts
const SUPPORTED_FILE_TYPES = {
  'image/png': 'png',
  'image/jpeg': 'jpg',
  'image/webp': 'webp'
}

// ...

const extension = SUPPORTED_FILE_TYPES[file.type]

if (!extension) {
  alert(`Files of type ${file.type} aren't supported. Files of type png, jpg, and webp are supported.`)
}
```

Reply:
If we still need `mimeTypes` or `extensions` as an array, we can do `Object.keys(SUPPORTED_FILE_TYPES)` or `Object.values(SUPPORTED_FILE_TYPES)`